### ISSUE TITLE
SqlServerProtocol: New resource proposal

### ISSUE THREAD

**Issue by johlju** [dsccommunity/SqlServerDsc#1377]
### Description
This resource should enable, disable and manage protocol properties for each protocol 'Named Pipes', 'Shared Memory' and 'TCP/IP'. 

For example this resource should manage the property `ListenOnAllIPs`, `KeepAlive` and `Enabled` which is protocol properties (see output below). 

```
PS > $managedComputerObject.ServerInstances['SQL2017'].ServerProtocols

Parent              : Microsoft.SqlServer.Management.Smo.Wmi.ServerInstance
DisplayName         : Named Pipes
HasMultiIPAddresses : False
IsEnabled           : False
IPAddresses         : {}
ProtocolProperties  : {Enabled, PipeName}
Urn                 : ManagedComputer[@Name='SQLTEST']/ServerInstance[@Name='SQL2017']/ServerProtocol[@Name='Np']
Name                : Np
Properties          : {Name=DisplayName/Type=System.String/Writable=False/Value=Named Pipes, Name=HasMultiIPAddresses/T
                      ype=System.Boolean/Writable=False/Value=False, Name=IsEnabled/Type=System.Boolean/Writable=True/V
                      alue=False}
UserData            :
State               : Existing

Parent              : Microsoft.SqlServer.Management.Smo.Wmi.ServerInstance
DisplayName         : Shared Memory
HasMultiIPAddresses : False
IsEnabled           : True
IPAddresses         : {}
ProtocolProperties  : {Enabled}
Urn                 : ManagedComputer[@Name='SQLTEST']/ServerInstance[@Name='SQL2017']/ServerProtocol[@Name='Sm']
Name                : Sm
Properties          : {Name=DisplayName/Type=System.String/Writable=False/Value=Shared Memory, Name=HasMultiIPAddresses
                      /Type=System.Boolean/Writable=False/Value=False, Name=IsEnabled/Type=System.Boolean/Writable=True
                      /Value=True}
UserData            :
State               : Existing

Parent              : Microsoft.SqlServer.Management.Smo.Wmi.ServerInstance
DisplayName         : TCP/IP
HasMultiIPAddresses : True
IsEnabled           : True
IPAddresses         : {IP1, IP2, IP3, IP4...}
ProtocolProperties  : {Enabled, KeepAlive, ListenOnAllIPs}
Urn                 : ManagedComputer[@Name='SQLTEST']/ServerInstance[@Name='SQL2017']/ServerProtocol[@Name='Tcp']
Name                : Tcp
Properties          : {Name=DisplayName/Type=System.String/Writable=False/Value=TCP/IP, Name=HasMultiIPAddresses/Type=S
                      ystem.Boolean/Writable=False/Value=True, Name=IsEnabled/Type=System.Boolean/Writable=True/Value=T
                      rue}
UserData            :
State               : Existing
```

### Proposed properties
```
[ClassVersion("1.0.0.0"), FriendlyName("SqlServerProtocol")]
class MSFT_SqlServerProtocol : OMI_BaseResource
{
    [Write, Description("The host name of the SQL Server to be configured. Default value is $env:COMPUTERNAME.")] String ServerName;
    [Key, Description("The name of the SQLf instance to be configured.")] String InstanceName;
    [Key, Description("The name of protocol to be configured."), ValueMap{"SharedMemory", "NamedPipes", "TcpIp"}, Values{"SharedMemory", "NamedPipes", "TcpIp"}] String ProtocolName;
    [Write, Description("Enables or disables the protocol.")] Boolean Enabled;
    [Write, Description("Specifies to listen on all IPs. Only used for the TCP/IP protocol.")] Boolean ListenOnAllIPs;
    [Write, Description("Specifies the keep alive duration. Only used for the TCP/IP protocol.")] UInt16 KeepAlive;
    [Write, Description("Specifies the named pipe name. Only used for the Named Pipes protocol.")] String PipeName;
    [Write, Description("If set to $true then SQL Server and dependent services will be restarted if a change to the configuration is made. The default value is $false.")] Boolean RestartService;
    [Write, Description("Timeout value for restarting the SQL Server services. The default value is 120 seconds.")] UInt16 RestartTimeout;
}
```
### Special considerations or limitations

This is a redesign of the resource `SqlServerNetwork` and splitting it up into two new resources will simplify the code and therefore easier to maintain. This resource could be added side-by-side with `SqlServerNetwork` and together with `SqlServerProtocolTcpIP` will deprecate the resource `SqlServerNetwork`.
This was discussed in closed PR https://github.com/PowerShell/SqlServerDsc/pull/1045.

**Comment by johlju** [comment]
There might be some code that can be reused in the closed PR https://github.com/PowerShell/SqlServerDsc/pull/1045.

**Comment by johlju** [comment]
Also this issue also closes issue #339.

**Comment by johlju** [comment]
Example code to verify and enable Named Pipes in this https://github.com/dsccommunity/SqlServerDsc/issues/1161#issuecomment-418921734.
